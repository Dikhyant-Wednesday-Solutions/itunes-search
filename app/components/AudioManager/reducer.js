/*
 *
 * AudioManager reducer
 *
 */
import { produce } from 'immer';
import { createActions } from 'reduxsauce';

export const { Types: audioManagerTypes, Creators: audioManagerCreators } = createActions({
  play: {},
  pause: {},
  playNew: ['src'],
  reset: {}
});

export const audioState = {
  PAUSED: 'PAUSED',
  PLAYING: 'PLAYING',
  ENDED: 'ENDED',
  NO_SONG: 'NO_SONG'
};

export const initialState = { src: null, audioState: audioState.NO_SONG };

export const audioManagerReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case audioManagerTypes.PLAY:
        draft.audioState = audioState.PLAYING;
        break;
      case audioManagerTypes.PAUSE:
        draft.audioState = audioState.PAUSED;
        break;
      case audioManagerTypes.PLAY_NEW:
        draft.audioState = audioState.PLAYING;
        draft.src = action.src;
        break;
      case audioManagerTypes.RESET:
        draft.audioState = audioState.NO_SONG;
        draft.src = null;
        break;
      default:
    }
  });

export default audioManagerReducer;
