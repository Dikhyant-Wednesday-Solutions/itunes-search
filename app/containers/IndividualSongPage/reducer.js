/*
 *
 * IndividualSongPage reducer
 *
 */
import { produce } from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const { Types: individualSongPageTypes, Creators: individualSongPageCreators } = createActions({
  requestGetItuneSong: ['songId'],
  successGetItuneSong: ['data'],
  failureGetItuneSong: ['error'],
  clearItuneSong: {}
});
export const initialState = { songId: null, songData: {}, songError: null, loading: null };

export const individualSongPageReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case individualSongPageTypes.REQUEST_GET_ITUNE_SONG:
        draft.songId = action.songId;
        draft.loading = true;
        break;
      case individualSongPageTypes.CLEAR_ITUNE_SONG:
        draft.songId = null;
        draft.songError = null;
        draft.songData = {};
        draft.loading = null;
        break;
      case individualSongPageTypes.SUCCESS_GET_ITUNE_SONG:
        draft.songData = action.data;
        draft.songError = null;
        draft.loading = false;
        break;
      case individualSongPageTypes.FAILURE_GET_ITUNE_SONG:
        draft.songError = get(action.error, 'message', 'something_went_wrong');
        draft.songData = null;
        draft.loading = false;
        break;
    }
  });

export default individualSongPageReducer;
