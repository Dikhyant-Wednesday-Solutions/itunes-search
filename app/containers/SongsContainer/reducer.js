/*
 *
 * SongsContainer reducer
 *
 */
import { produce } from 'immer';
import { createActions } from 'reduxsauce';
import get from 'lodash/get';

export const { Types: songsContainerTypes, Creators: songsContainerCreators } = createActions({
  requestGetItuneSongs: ['songName'],
  successGetItuneSongs: ['data'],
  failureGetItuneSongs: ['error'],
  clearItuneSongs: {}
});
export const initialState = { songName: null, songsData: {}, songsError: null, loading: null };

export const songsContainerReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case songsContainerTypes.REQUEST_GET_ITUNE_SONGS:
        draft.songName = action.songName;
        draft.loading = true;
        break;
      case songsContainerTypes.CLEAR_ITUNE_SONGS:
        draft.songName = null;
        draft.songsError = null;
        draft.songsData = {};
        draft.loading = null;
        break;
      case songsContainerTypes.SUCCESS_GET_ITUNE_SONGS:
        draft.songsData = action.data;
        draft.songsError = null;
        draft.loading = false;
        break;
      case songsContainerTypes.FAILURE_GET_ITUNE_SONGS:
        draft.songsError = get(action.error, 'message', 'something_went_wrong');
        draft.songsData = null;
        draft.loading = false;
        break;
    }
  });

export default songsContainerReducer;
