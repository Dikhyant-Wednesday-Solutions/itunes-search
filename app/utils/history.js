import { createBrowserHistory } from 'history';
import { matchPath } from 'react-router';
import routeConstants from './routeConstants';

const routes = Object.keys(routeConstants);

export function findCommonRoutePrefix(routeArr) {
  let maxCommon = routeArr[0];
  for (let i = 1; i < routeArr.length; i++) {
    stringLoop: for (let j = 0; j < maxCommon.length; j++) {
      if (!routeArr[i][j]) {
        maxCommon = routeArr[i];
        break stringLoop;
      } else if (maxCommon[j] !== routeArr[i][j]) {
        maxCommon = routeArr[i].substring(0, j);
        break stringLoop;
      }
    }
  }
  return maxCommon;
}

/** @param {string} pathname */
export function getBaseUrl(pathname) {
  let baseURL = '';

  if (process.env.ENVIRONMENT_NAME === 'production') {
    baseURL = '/react-template';
  }

  if (process.env.ENVIRONMENT_NAME === 'uat') {
    let baseURLCollection = [];

    routes.forEach((routeKey) => {
      const route = routeConstants[routeKey].route;

      if (pathname.includes(route)) {
        // feat/repos /repos
        if (pathname.substring(pathname.length - route.length, pathname.length) === route) {
          baseURLCollection.push(pathname.substring(0, pathname.length - route.length));
        }
        if (pathname.substring(pathname.length - route.length - 1, pathname.length) === `${route}/`) {
          baseURLCollection.push(pathname.substring(0, pathname.length - route.length - 1));
        }
      } else {
        if (route.includes(':')) {
          const regex = /^(?!:)\/[\w]+/;
          const matches = regex.exec(route);

          let matchLastIndex = pathname.lastIndexOf(matches[0]);
          const pathToMatch = pathname.substring(matchLastIndex);
          const isMatch = matchPath(pathToMatch, {
            path: route,
            exact: true
          });
          if (isMatch) {
            baseURLCollection.push(pathname.substring(0, matchLastIndex));
          }
        }
      }
    });
    if (!baseURLCollection.length) {
      baseURL = pathname;
    } else {
      baseURL = findCommonRoutePrefix(baseURLCollection);
    }
  }
  return baseURL;
}

const history = createBrowserHistory({ basename: getBaseUrl(window.location.pathname) });
export default history;
