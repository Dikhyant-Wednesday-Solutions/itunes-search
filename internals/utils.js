function isUAT() {
  return process.env.ENVIRONMENT_NAME === 'uat';
}

function isProd() {
  return process.env.ENVIRONMENT_NAME === 'production';
}

function getBranchName() {
  return process.env.BRANCH_NAME ? `/${process.env.BRANCH_NAME}/` : './';
}

function getBasePublicPath() {
  return isUAT() ? getBranchName() : isProd() ? '/react-template' : '/';
}

module.exports = { getBasePublicPath, isUAT, isProd };
