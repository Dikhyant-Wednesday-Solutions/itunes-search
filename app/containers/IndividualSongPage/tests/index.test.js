import React from 'react';
import Router from "react-router-dom";
import { i18n } from '@lingui/core';
import { renderProvider, timeout } from '@utils/testUtils';
import { translate } from '@app/utils';
import en from "@app/translations/en.json";
import { IndividualSongPage, mapDispatchToProps } from '../index';
import songData from './song.data.json';
import { individualSongPageTypes } from '../reducer';

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useParams: jest.fn(),
    useLocation: jest.fn(),
}));

i18n.load('en', en);
i18n.activate('en');

describe('<IndividualSongPage /> test suite', () => {
    beforeEach(() => {
        const songId = "1440833081";
        jest.spyOn(Router, 'useParams').mockReturnValue({ id: songId });
        jest.spyOn(Router, 'useLocation').mockReturnValue({state: {}});
    });
    it('should render and match snapshot', () => {
        const { baseElement } = renderProvider(<IndividualSongPage/>);
        expect(baseElement).toMatchSnapshot();
    }); 

    it('should call dispatchRequestGetItuneSong when no song info is provided by previous page and song id is available in url param', () => {
        const songId = "1440833081";
        const dispatchRequestGetItuneSongSpy = jest.fn();
        console.log("dispatchRequestGetItuneSongSpy instanceof Function = ", dispatchRequestGetItuneSongSpy instanceof Function);
        renderProvider(<IndividualSongPage dispatchRequestGetItuneSong={dispatchRequestGetItuneSongSpy}/>);
        expect(dispatchRequestGetItuneSongSpy).toHaveBeenCalledWith(songId);
    });

    it('should not call dispatchRequestGetItuneSong when song info from previous page is provided', () => {
        jest.spyOn(Router, 'useLocation').mockReturnValue({state: {songData: songData}});
        const dispatchRequestGetItuneSongSpy = jest.fn();
        renderProvider(<IndividualSongPage dispatchRequestGetItuneSong={dispatchRequestGetItuneSongSpy}/>);
        expect(dispatchRequestGetItuneSongSpy).not.toHaveBeenCalled();
    });

    it('should render error component when songError is provided', () => {
        const defaultError = translate('something_went_wrong');
        console.log(defaultError);
        const { getByTestId } = renderProvider(<IndividualSongPage songError={defaultError} />);
        expect(getByTestId('error-message')).toBeInTheDocument();
        expect(getByTestId('error-message')).toHaveTextContent(defaultError);
    });

    it('should render Skeleton Comp when "loading" is true', async () => {
        const { getAllByTestId } = renderProvider(
          <IndividualSongPage loading={true} />
        );
        expect(getAllByTestId('skeleton').length).toBe(3);
    });

    it('should validate mapDispatchToProps actions', async () => {
        const dispatchIndieSongSpy = jest.fn();
        const songId = "1440833081";
        const actions = {
            dispatchRequestGetItuneSong: {
                type: individualSongPageTypes.REQUEST_GET_ITUNE_SONG,
                songId
            },
            dispatchClearItuneSong: {
                type: individualSongPageTypes.CLEAR_ITUNE_SONG,
            }
        }

        const props = mapDispatchToProps(dispatchIndieSongSpy);
        props.dispatchRequestGetItuneSong(songId);
        expect(dispatchIndieSongSpy).toHaveBeenCalledWith(actions.dispatchRequestGetItuneSong);
        await timeout(500);

        props.dispatchClearItuneSong();
        expect(dispatchIndieSongSpy).toHaveBeenCalledWith(actions.dispatchClearItuneSong);
    })
})