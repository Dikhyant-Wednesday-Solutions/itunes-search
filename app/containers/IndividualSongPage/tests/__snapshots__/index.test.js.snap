// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IndividualSongPage /> test suite should render and match snapshot 1`] = `
<body>
  <div>
    <div
      class="css-zjik7 e193px3m4"
    >
      <audio>
        <source
          type="type/mpeg"
        />
      </audio>
    </div>
  </div>
</body>
`;
